In [1]:
import pandas as pd

df = pd.read_csv("synthetic_logs.csv")
df.head()

timestamp           source  \
0  2025-06-27 07:20:25        ModernCRM   
1      1/14/2025 23:07        ModernCRM   
2       1/17/2025 1:29  AnalyticsEngine   
3  2025-07-12 00:24:16         ModernHR   
4  2025-06-02 18:25:23    BillingSystem   

                                         log_message    target_label  \
0  nova.osapi_compute.wsgi.server [req-b9718cd8-f...     HTTP Status   
1     Email service experiencing issues with sending  Critical Error   
2          Unauthorized access to data was attempted  Security Alert   
3  nova.osapi_compute.wsgi.server [req-4895c258-b...     HTTP Status   
4  nova.osapi_compute.wsgi.server [req-ee8bc8ba-9...     HTTP Status   

  complexity  
0       bert  
1       bert  
2       bert  
3       bert  
4       bert

In [3]:
df.shape

(2410, 5)

In [2]:
df["source"].value_counts()

source
ThirdPartyAPI      496
ModernHR           492
BillingSystem      479
AnalyticsEngine    471
ModernCRM          465
LegacyCRM            7
Name: count, dtype: int64

In [8]:
df["target_label"].value_counts()

target_label
HTTP Status            1017
Security Alert          371
System Notification     356
Error                   177
Resource Usage          177
Critical Error          161
User Action             144
Workflow Error            4
Deprecation Warning       3
Name: count, dtype: int64

In [4]:
df["complexity"].value_counts()

complexity
bert     1903
regex     500
llm         7
Name: count, dtype: int64

In [5]:
df.isnull().sum()

timestamp       0
source          0
log_message     0
target_label    0
complexity      0
dtype: int64

In [6]:
df[df.target_label=='System Notification'].sample(10)

timestamp           source  \
1948  5/14/2025 18:04        ModernCRM   
2043  9/12/2025 20:20    ThirdPartyAPI   
2078   4/9/2025 17:00  AnalyticsEngine   
1834   6/13/2025 5:43        ModernCRM   
1033  7/26/2025 13:27         ModernHR   
2376   6/27/2025 8:47        ModernCRM   
191    5/6/2025 23:58    ThirdPartyAPI   
728   5/18/2025 18:19    BillingSystem   
1207   1/9/2025 22:57        ModernCRM   
762     9/5/2025 3:51  AnalyticsEngine   

                                            log_message         target_label  \
1948             Backup started at 2025-11-23 01:05:16.  System Notification   
2043           System reboot initiated by user User262.  System Notification   
2078  File data_4759.csv uploaded successfully by us...  System Notification   
1834               Backup ended at 2025-02-21 02:50:57.  System Notification   
1033             Backup started at 2025-02-12 06:47:30.  System Notification   
2376                   System updated to version 2.0.5.  System Notification   
191              Backup started at 2025-12-09 10:19:11.  System Notification   
728                Backup ended at 2025-08-06 22:28:29.  System Notification   
1207  File data_3277.csv uploaded successfully by us...  System Notification   
762                    System updated to version 5.0.6.  System Notification   

     complexity  
1948      regex  
2043      regex  
2078      regex  
1834      regex  
1033      regex  
2376      regex  
191       regex  
728       regex  
1207      regex  
762       regex

In [9]:
df[df.log_message.str.startswith("System reboot initiated by user")].sample(10)

timestamp           source  \
307    4/12/2025 0:41    BillingSystem   
1865  5/11/2025 10:58  AnalyticsEngine   
2253   10/7/2025 2:20         ModernHR   
1524  11/30/2025 2:39    ThirdPartyAPI   
800   8/15/2025 12:14    BillingSystem   
2360    5/1/2025 4:21    ThirdPartyAPI   
1776  2/21/2025 11:56         ModernHR   
852    3/31/2025 5:20        ModernCRM   
1304   8/10/2025 6:18    ThirdPartyAPI   
140    9/11/2025 8:49  AnalyticsEngine   

                                   log_message         target_label complexity  
307   System reboot initiated by user User929.  System Notification      regex  
1865  System reboot initiated by user User932.  System Notification      regex  
2253  System reboot initiated by user User644.  System Notification      regex  
1524  System reboot initiated by user User278.  System Notification      regex  
800   System reboot initiated by user User901.  System Notification      regex  
2360  System reboot initiated by user User876.  System Notification      regex  
1776  System reboot initiated by user User155.  System Notification      regex  
852   System reboot initiated by user User811.  System Notification      regex  
1304  System reboot initiated by user User758.  System Notification      regex  
140   System reboot initiated by user User639.  System Notification      regex

# clustering

In [ ]:
from sklearn.cluster import DBSCAN
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('all-MiniLM-L6-v2')
embeddings = model.encode(df['log_message'].tolist())
embeddings[:2]

In [11]:
clustering = DBSCAN(eps=0.2, min_samples=1, metric='cosine').fit(embeddings)
df['cluster'] = clustering.labels_
df.head()

timestamp           source  \
0  2025-06-27 07:20:25        ModernCRM   
1      1/14/2025 23:07        ModernCRM   
2       1/17/2025 1:29  AnalyticsEngine   
3  2025-07-12 00:24:16         ModernHR   
4  2025-06-02 18:25:23    BillingSystem   

                                         log_message    target_label  \
0  nova.osapi_compute.wsgi.server [req-b9718cd8-f...     HTTP Status   
1     Email service experiencing issues with sending  Critical Error   
2          Unauthorized access to data was attempted  Security Alert   
3  nova.osapi_compute.wsgi.server [req-4895c258-b...     HTTP Status   
4  nova.osapi_compute.wsgi.server [req-ee8bc8ba-9...     HTTP Status   

  complexity  cluster  
0       bert        0  
1       bert        1  
2       bert        2  
3       bert        0  
4       bert        0

In [12]:
clusters = df.groupby('cluster')['log_message'].apply(list)
sorted_clusters = clusters.sort_values(key=lambda x: x.map(len), ascending=False)
print("Clustered Patterns:")
for cluster_id, messages in sorted_clusters.items():
    if len(messages) > 10:
        print(f"Cluster {cluster_id}:")
        for msg in messages[:5]:
            print(f"  {msg}")

Clustered Patterns:
Cluster 0:
  nova.osapi_compute.wsgi.server [req-b9718cd8-f65e-49cc-8349-6cf7122af137 113d3a99c3da401fbd62cc2caa5b96d2 54fadb412c4e40cdbaed9335e4c35a9e - - -] 10.11.10.1 "GET /v2/54fadb412c4e40cdbaed9335e4c35a9e/servers/detail HTTP/1.1" status: 200 len: 1893 time: 0.2675118
  nova.osapi_compute.wsgi.server [req-4895c258-b2f8-488f-a2a3-4fae63982e48 113d3a99c3da401fbd62cc2caa5b96d2 54fadb412c4e40cdbaed9335e4c35a9e - - -] 10.11.10.1 "GET /v2/54fadb412c4e40cdbaed9335e4c35a9e/servers/detail HTTP/1.1" HTTP status code -  200 len: 211 time: 0.0968180
  nova.osapi_compute.wsgi.server [req-ee8bc8ba-9265-4280-9215-dbe000a41209 113d3a99c3da401fbd62cc2caa5b96d2 54fadb412c4e40cdbaed9335e4c35a9e - - -] 10.11.10.1 "GET /v2/54fadb412c4e40cdbaed9335e4c35a9e/servers/detail HTTP/1.1" RCODE  200 len: 1874 time: 0.2280791
  nova.osapi_compute.wsgi.server [req-f0bffbc3-5ab0-4916-91c1-0a61dd7d4ec2 113d3a99c3da401fbd62cc2caa5b96d2 54fadb412c4e40cdbaed9335e4c35a9e - - -] 10.11.10.1 "GET /v2

In [15]:
df[df["cluster"]==53]

timestamp           source  \
283    1/25/2025 13:45        ModernCRM   
344   11/13/2025 13:20    BillingSystem   
378   12/24/2025 19:10  AnalyticsEngine   
454    5/20/2025 11:01  AnalyticsEngine   
530   12/12/2025 13:41        ModernCRM   
620     1/19/2025 2:53    BillingSystem   
645      8/8/2025 0:27         ModernHR   
744    11/6/2025 15:30    ThirdPartyAPI   
1110    6/1/2025 15:58    BillingSystem   
1239     4/9/2025 8:41         ModernHR   
1302    2/12/2025 3:21         ModernHR   
1382  12/27/2025 11:59    BillingSystem   
1569   2/21/2025 21:16         ModernHR   
1644   6/20/2025 19:54  AnalyticsEngine   
1936   12/2/2025 20:50    BillingSystem   
1977   2/28/2025 11:26  AnalyticsEngine   
2054    5/16/2025 8:58        ModernCRM   
2209    3/30/2025 8:54        ModernCRM   
2279   10/10/2025 3:59  AnalyticsEngine   
2299    7/10/2025 5:36    BillingSystem   

                                            log_message target_label  \
283   Module X experienced an invalid data format issue        Error   
344          Input format mismatch occurred in module X        Error   
378   Module X reported an error in input format val...        Error   
454   Module X failed to process input due to format...        Error   
530   Input data format in module X was invalid or c...        Error   
620   Input to module X did not conform to expected ...        Error   
645   Format of input data in module X was not recog...        Error   
744   Error in input format occurred during module X...        Error   
1110       Invalid input format caused module X to fail        Error   
1239   Input to module X was not in the expected format        Error   
1302      Module X failed to validate input data format        Error   
1382     Module X experienced a format validation error        Error   
1569             Invalid input syntax found in module X        Error   
1644  Module X encountered an issue with improperly ...        Error   
1936  Module X input processing failed due to invali...        Error   
1977  Invalid data encountered during execution of m...        Error   
2054  Module X reported an invalid input format exce...        Error   
2209               Malformed input detected in module X        Error   
2279  Module X was unable to parse input due to form...        Error   
2299  Error occurred due to incorrect input in module X        Error   

     complexity  cluster  
283        bert       53  
344        bert       53  
378        bert       53  
454        bert       53  
530        bert       53  
620        bert       53  
645        bert       53  
744        bert       53  
1110       bert       53  
1239       bert       53  
1302       bert       53  
1382       bert       53  
1569       bert       53  
1644       bert       53  
1936       bert       53  
1977       bert       53  
2054       bert       53  
2209       bert       53  
2279       bert       53  
2299       bert       53

# Classfication Using Regex

In [16]:
import re
def classify_with_regex(log_message):
    regex_patterns = {
        r"User User\d+ logged (in|out).": "User Action",
        r"Backup (started|ended) at .*": "System Notification",
        r"Backup completed successfully.": "System Notification",
        r"System updated to version .*": "System Notification",
        r"File .* uploaded successfully by user .*": "System Notification",
        r"Disk cleanup completed successfully.": "System Notification",
        r"System reboot initiated by user .*": "System Notification",
        r"Account with ID .* created by .*": "User Action"
    }
    for pattern, label in regex_patterns.items():
        if re.search(pattern, log_message):
            return label
    return None

## Examples

In [17]:
classify_with_regex("User User123 logged in.")

'User Action'

In [18]:
classify_with_regex("System reboot initiated by user User179.")

'System Notification'

In [19]:
classify_with_regex("Hey you, chill bro")

In [20]:
df['regex_label'] = df['log_message'].apply(lambda x: classify_with_regex(x))
df[df['regex_label'].notnull()]

timestamp         source  \
7     10/11/2025 8:44       ModernHR   
14      1/4/2025 1:43  ThirdPartyAPI   
15      5/1/2025 9:41      ModernCRM   
18    2/22/2025 17:49      ModernCRM   
27    9/24/2025 19:57  ThirdPartyAPI   
...               ...            ...   
2376   6/27/2025 8:47      ModernCRM   
2381    9/5/2025 6:39  ThirdPartyAPI   
2394   4/3/2025 13:13       ModernHR   
2395   5/2/2025 14:29  ThirdPartyAPI   
2403   10/1/2025 1:31      ModernCRM   

                                            log_message         target_label  \
7     File data_6169.csv uploaded successfully by us...  System Notification   
14    File data_3847.csv uploaded successfully by us...  System Notification   
15                       Backup completed successfully.  System Notification   
18             Account with ID 5351 created by User634.          User Action   
27                             User User685 logged out.          User Action   
...                                                 ...                  ...   
2376                   System updated to version 2.0.5.  System Notification   
2381               Disk cleanup completed successfully.  System Notification   
2394               Disk cleanup completed successfully.  System Notification   
2395               Backup ended at 2025-05-06 11:23:16.  System Notification   
2403                     Backup completed successfully.  System Notification   

     complexity  cluster          regex_label  
7         regex        4  System Notification  
14        regex        4  System Notification  
15        regex        8  System Notification  
18        regex        9          User Action  
27        regex       11          User Action  
...         ...      ...                  ...  
2376      regex       21  System Notification  
2381      regex       32  System Notification  
2394      regex       32  System Notification  
2395      regex       13  System Notification  
2403      regex        8  System Notification  

[500 rows x 7 columns]

In [21]:
df[df['regex_label'].isnull()].shape

(1910, 7)

In [23]:
df.shape

(2410, 7)

In [22]:
1910+500

2410

### Out of 2410 rows
1. 500 rows are classfied using regex function
2. 1910 rows couldn`t be classified. Now we will train to classify these rows using BERT and LLM

# Classfication using BERT

In [24]:
df_non_regex = df[df['regex_label'].isnull()].copy()
df_non_regex.shape

(1910, 7)

In [25]:
df_non_regex.head()

timestamp           source  \
0  2025-06-27 07:20:25        ModernCRM   
1      1/14/2025 23:07        ModernCRM   
2       1/17/2025 1:29  AnalyticsEngine   
3  2025-07-12 00:24:16         ModernHR   
4  2025-06-02 18:25:23    BillingSystem   

                                         log_message    target_label  \
0  nova.osapi_compute.wsgi.server [req-b9718cd8-f...     HTTP Status   
1     Email service experiencing issues with sending  Critical Error   
2          Unauthorized access to data was attempted  Security Alert   
3  nova.osapi_compute.wsgi.server [req-4895c258-b...     HTTP Status   
4  nova.osapi_compute.wsgi.server [req-ee8bc8ba-9...     HTTP Status   

  complexity  cluster regex_label  
0       bert        0        None  
1       bert        1        None  
2       bert        2        None  
3       bert        0        None  
4       bert        0        None

In [26]:
df_non_regex['source'].value_counts()

source
ModernHR           402
ThirdPartyAPI      386
ModernCRM          373
AnalyticsEngine    371
BillingSystem      371
LegacyCRM            7
Name: count, dtype: int64

## We can clearly see that legacyCRM has less value for training model. We use LLM for this source to predict the classification.

In [27]:
df_legacy = df_non_regex[df_non_regex.source=="LegacyCRM"]
df_legacy

timestamp     source  \
60    2025-10-06 16:55:23  LegacyCRM   
255   2025-05-03 16:55:35  LegacyCRM   
377   2025-06-24 12:16:29  LegacyCRM   
1325  2025-04-17 07:33:44  LegacyCRM   
1734  2025-04-30 07:47:30  LegacyCRM   
1826  2025-01-23 10:33:36  LegacyCRM   
2217  2025-05-12 09:46:54  LegacyCRM   

                                            log_message         target_label  \
60    Lead conversion failed for prospect ID 7842 du...       Workflow Error   
255   API endpoint 'getCustomerDetails' is deprecate...  Deprecation Warning   
377   Customer follow-up process for lead ID 5621 fa...       Workflow Error   
1325  Escalation rule execution failed for ticket ID...       Workflow Error   
1734  The 'ExportToCSV' feature is outdated. Please ...  Deprecation Warning   
1826  Support for legacy authentication methods will...  Deprecation Warning   
2217  Task assignment for TeamID 3425 could not comp...       Workflow Error   

     complexity  cluster regex_label  
60          llm       24        None  
255         llm       48        None  
377         llm       62        None  
1325        llm      105        None  
1734        llm      118        None  
1826        llm      122        None  
2217        llm      133        None

In [28]:
df_non_legacy = df_non_regex[df_non_regex.source!="LegacyCRM"]
df_non_legacy

timestamp           source  \
0     2025-06-27 07:20:25        ModernCRM   
1         1/14/2025 23:07        ModernCRM   
2          1/17/2025 1:29  AnalyticsEngine   
3     2025-07-12 00:24:16         ModernHR   
4     2025-06-02 18:25:23    BillingSystem   
...                   ...              ...   
2405  2025-08-13 07:29:25         ModernHR   
2406       1/11/2025 5:32         ModernHR   
2407  2025-08-03 03:07:47    ThirdPartyAPI   
2408     11/11/2025 11:52    BillingSystem   
2409     12/25/2025 13:21  AnalyticsEngine   

                                            log_message    target_label  \
0     nova.osapi_compute.wsgi.server [req-b9718cd8-f...     HTTP Status   
1        Email service experiencing issues with sending  Critical Error   
2             Unauthorized access to data was attempted  Security Alert   
3     nova.osapi_compute.wsgi.server [req-4895c258-b...     HTTP Status   
4     nova.osapi_compute.wsgi.server [req-ee8bc8ba-9...     HTTP Status   
...                                                 ...             ...   
2405  nova.osapi_compute.wsgi.server [req-96c3ec98-2...     HTTP Status   
2406  User 3844 account experienced multiple failed ...  Security Alert   
2407  nova.metadata.wsgi.server [req-b6d4a270-accb-4...     HTTP Status   
2408      Email service affected by failed transmission  Critical Error   
2409  Repeated failed login attempts occurred for us...  Security Alert   

     complexity  cluster regex_label  
0          bert        0        None  
1          bert        1        None  
2          bert        2        None  
3          bert        0        None  
4          bert        0        None  
...         ...      ...         ...  
2405       bert        0        None  
2406       bert        7        None  
2407       bert        0        None  
2408       bert        1        None  
2409       bert        7        None  

[1903 rows x 7 columns]

In [29]:
model = SentenceTransformer('all-MiniLM-L6-v2')  # Lightweight embedding model
embeddings_filtered = model.encode(df_non_legacy['log_message'].tolist())

In [30]:
X = embeddings_filtered
y = df_non_legacy['target_label'].values

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)

                precision    recall  f1-score   support

Critical Error       0.91      1.00      0.95        48
         Error       0.98      0.89      0.93        47
   HTTP Status       1.00      1.00      1.00       304
Resource Usage       1.00      1.00      1.00        49
Security Alert       1.00      0.99      1.00       123

      accuracy                           0.99       571
     macro avg       0.98      0.98      0.98       571
  weighted avg       0.99      0.99      0.99       571



In [33]:
import joblib
joblib.dump(clf,"log_classifier.joblib")

['log_classifier.joblib']